src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -516,6 +516,66 @@ class OrderedRVD(
 
   override def toUnpartitionedRVD: UnpartitionedRVD =
     new UnpartitionedRVD(typ.rowType, crdd)
+
+  def orderedLeftJoinDistinctAndInsert(
+    right: OrderedRVD,
+    root: String,
+    lift: Option[String] = None,
+    dropLeft: Option[Array[String]] = None
+  ): OrderedRVD = {
+    assert(!typ.key.contains(root))
+
+    val valueStruct = right.typ.valueType
+    val rightRowType = right.typ.rowType
+    val liftField = lift.map { name => rightRowType.field(name) }
+    val valueType = liftField.map(_.typ).getOrElse(valueStruct)
+
+    val removeLeft = dropLeft.map(_.toSet).getOrElse(Set.empty)
+    val keepIndices = rowType.fields.filter(f => !removeLeft.contains(f.name)).map(_.index).toArray
+    val newRowType = TStruct(
+      keepIndices.map(i => rowType.fieldNames(i) -> rowType.types(i)) ++ Array(root -> valueType): _*)
```

Comment:
we should force the value type to be optional, right?

Reply:
good call.